tools/create_service_diagnostics_bundle.sh

```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/env bash
+
+set -eu -o pipefail
+
+function usage () {
+  echo 'Usage: ./create_service_diagnostics_bundle.sh <package name> <service name> [<proceed>]'
+  echo
+  echo 'Example: ./create_service_diagnostics_bundle.sh cassandra /prod/cassandra yes'
+}
+
+if [ "${#}" -lt 2 ] || [ "${#}" -gt 3 ]; then
+  echo -e "create_service_diagnostics_bundle.sh takes either 2 or 3 arguments but was given ${#}\\n"
+  usage
+  exit 1
+fi
+
+readonly REQUIREMENTS='docker'
+
+for requirement in ${REQUIREMENTS}; do
+  if ! [[ -x $(command -v "${requirement}") ]]; then
+    echo "You need to install '${requirement}' to run this script"
```

Comment:
I weighted the short-term implementation alternatives (detailed in the design document) and found that Docker was the most appealing, even with its shortcomings.

The long-term implementation should be what @elezar recommended:
> What about a `dcos` subcommand that targets services? Remember that most of our services (except Kafka and Spark at the moment) use the same CLI. We could have a "service-diag" CLI command that is installed and can target any service:
> dcos package install --yes --cli service-diagnostics
> dcos service-diagnostics collect --name=/my/elastic/service
>
> Existing examples of such CLIs are:
> * The enterprise CLI
> * The tunnel CLI

With this implementation we just want to provide customers something that is better than the current situation, which they can always fall-back on in case of problems running this script.

Reply:
It might be most appealing to us in the short term, but I'm afraid we'll just have to add a tarball form of distribution soon, once we hit a customer who cannot use docker.

AFAICT the only difference between using a tarball and the docker image found in https://docs.google.com/document/d/1STxZPbV5OJ8D-jj1u72jKLlTKq3FsG2nniER3_qWNr8/edit#heading=h.g95cohdwgsdd is that "the environment might be unknown" which seems very vague to me. OTOH a customer not being to use docker is a very concrete limitation.

I'm afraid we'll have to fix the "Python environment (dependencies, versions, ENV, etc.) management is very bad" case either way - whether we do it in a tarballed directory with  venv+pip or in a docker image with system-level pip.

Or am I missing something?

Reply:
> I'm afraid we'll have to fix the "Python environment (dependencies, versions, ENV, etc.) management is very bad" case either way - whether we do it in a tarballed directory with venv+pip or in a docker image with system-level pip.

That's true. In the Docker image that's already "fixed" so there's nothing to do :)

What about this: we release this using a Docker image and if some customer (and by "customer" here I'm assuming operations engineers or developers, the people who reach out to support and which I'd expect to have Docker installed in their system) can't use Docker *then* we work on figuring out how to distribute it with a tarball containing the whole controlled environment?

Reply:
We're actually far from "nothing to do" even in the docker case. Not even all direct dependencies are pinned to a fixed version, never mind indirect dependencies. We had a CI outage not so long ago exactly because of this problem.

So whenever we rebuild the docker image we will pick up new versions of some potentially broken dependencies - I had filed DCOS-41913 about this.

I also only now realized that you are reusing the `mesosphere/dcos-commons` docker image (which we use for CI) for this script. Additionally the script you added has no integration tests, so nobody will even notice if they break the docker image for the purpose of running this script, as long as it keeps working for CI.

Do you think I could convince you to add a simple integration test that runs this script against any of our services just to make sure it does not crash and produces *something*?

Reply:
> Right. Since the script runs off a versioned image (e.g. mpereira/dcos-commons:diagnostics-v0.1.0 (and not off a mutable tag like latest), unless the tag is force pushed the script environment will remain unchanged for its version.

That's very good, but I had a different scenario in mind: once we fix even a typo in the diagnostic script, we need to rebuild the docker image, which means we will pick up any changes that happened to any unpinned dependency we use.